### PR TITLE
Doc for custom_gradient to compute gradients for trainable params and computing over batch

### LINKED ISSUE #26270: [TF 2.0 API Docs] tf.custom_gradient
<em>Please make sure that this is a documentation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:doc_template</em>

**System information**
- TensorFlow version: 1.13.1
- Doc Link: https://www.tensorflow.org/api_docs/python/tf/custom_gradient

I've been having some difficulty using `custom_gradient` recently. Before implementing the op I actually have in mind, I'm trying to make a [simple polynomial op](https://gist.github.com/tsbertalan/b6c02bf6e39116d8446faa0159a011af) with `custom_gradient` in an effort to get my head around the requirements.. The documentation is not at all clear on how summing over batch and output indices should be done, and (until recently) wasn't clear on whether we should return the full *Jacobian* of the operation, or just the vector-Jacobian-product ("VJP") function. (I did notice some explanation of the latter issue made it in in the 1.12 -> 1.13 update, which was helpful).

I think this problem applies both the derivative WRT the op's input, and WRT to the parameters (called `grad_xs` and `grad_vars` in the documentation). However, when I incorporated the VJP tidbit in my sample code, I got correct results for the former, but not the latter, I believe because the implicit sum over batch was already present in the supplied `grad_ys`, and because the VJP also collapses out the index across `grad_ys`.

So, the lack of clarity about sums (accidentally) only hit me when trying to write the `grad_vars` part. In my [gist linked above](https://gist.github.com/tsbertalan/b6c02bf6e39116d8446faa0159a011af), I had to use a `reduce_sum` when computing $dy/dp$, which, for a polynomial, is the corresponding powers of $x$. (Since $dy/dx$ doesn't depend on $x$, this issue didn't appear when writing the `grad_xs` part.)

In addition to simply saying what you mean in this documentation (the sum of the gradient over examples is not "the gradient"--only the per-example gradient truly deserves that name! But I suppose that particular fight is a lost cause.), it would be good if there were at least some examples that exercised the `grad_vars` part and the `variables=None` keyword argument.

### PR DESCRIPTION
**Pull request by ashutosh1919**
Fixes #26270 .

Thanks to @tsbertalan for gist. I took help of that and soon will be raising PR to `tensorflow/examples` repo for adding the `custom_gradient` example for polynomial.

cc @alextp , @dynamicwebpaige , @tsbertalan .

### CODE REVIEW

**Comment by ashutosh1919** [comment]
The build failure is not because of the changes in this PR.
It because of bazel build failure and the error is something like this:
`'@llvm-project//llvm:AMDGPUCodeGen' is duplicated in the 'deps' attribute of rule 'llvm_gpu_backend'`

File: tensorflow/python/ops/custom_gradient.py [hunk 1]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
...
+    numpy=array([1., 1., 1.], dtype=float32)>
+  >>> tape.gradient(poly, trainable[0])  # gradient w.r.t weights dy/dweights
+  <tf.Tensor: shape=(2,),
+    dtype=float32,
+    numpy=array([6., 3.], dtype=float32)>
+  ```
+
+  Above example illustrates two things:
+  - Usage of trainable variable `weights` and computing it's gradient.
+  - Computing gradient in case of batches.
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:233]: Not sure what "Computing gradient in case of batches" means.

Maybe just delete this bit, but otherwise as a nit s/"computing it's gradient"/"computing its gradient"/

File: tensorflow/python/ops/custom_gradient.py [hunk 2]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
...
+
+  In example, You can see the inner `grad_fn` that accepts extra `variables`
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:235]: grammar/style nits: "In the example, the inner `grad_fn` accepts an extra `variables` input parameter and also returns an extra `grad_vars` output." (I'd leave off the "scalar gradient functions" bit, since it's stateless vs. stateful more than scalar vs. non-scalar)

File: tensorflow/python/ops/custom_gradient.py [hunk 3]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
...
+  input parameter and also returns extra `grad_vars` output as compared to
+  conventional scalar gradient functions. That second parameter is passed if
+  you have any trainable parameters in model and you need to compute the
+  gradient w.r.t. each of those variables and output it as list `grad_vars`.
+  Note here that default value of `variables` is set to `None` because the
+  gradient function should also be able to calculate gradient of scalar when
+  no `variables` are passed.
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:242]: "That extra argument is passed if you have any trainable parameters in your model. You need to compute the gradient w.r.t. each of those variables and output it as a list of `grad_vars`. Note here that default value of `variables` is set to `None` when no variables are used in the forward function."

File: tensorflow/python/ops/custom_gradient.py [hunk 4]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
...
+
+  You can see that the shape of `x` is (3,) and `weights` is (2,). Here, you
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:244]: Is this paragraph relevant to `tf.custom_gradient`? The docstring is already quite long, if we can cut this it would be helpful.

**Review by allenlavoie — COMMENTED** [review]
Thank you for writing this example! Overall it looks good, just some nits.

File: tensorflow/python/ops/custom_gradient.py [hunk 5]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
   Additional arguments to the inner `@tf.custom_gradient`-decorated function
   control the expected return values of the innermost function.
 
+  Above examples illustrates how you can create custom scalar gradient
```
[the remark this answers is not in the document]
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

File: tensorflow/python/ops/custom_gradient.py [hunk 6]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
...
+  ...   # Creating polynomial
+  ...   poly = weights[1] * x + weights[0]
+  ...
+  ...   def grad_fn(dpoly, variables=None):
+  ...     # dy/dx = weights[1] and we need to left multiply dpoly
+  ...     grad_xs = dpoly * weights[1]  # Scalar gradient
+  ...
+  ...     input_shape = len(x.shape)  # x contains how many dimensions
+  ...     grad_vars = []  # To store gradients of passed variables
+  ...     if variables is not None:
```
[the remark this answers is not in the document]
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py:186]: done

File: tensorflow/python/ops/custom_gradient.py [hunk 7]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
...
+  ...                 tf.reduce_sum(dy_dw_concat, axis=1)
+  ...             )
+  ...           else:
+  ...             # No need to reduce if x is scalar
+  ...             grad_vars.append(dy_dw_concat)
+  ...         else:  # Create elif if you have more variables
+  ...           grad_vars.append(None)
+  ...     return grad_xs, grad_vars
+  ...   return poly, grad_fn
+  >>> x = tf.Variable([1., 2., 3.])
```
[the remark this answers is not in the document]
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

File: tensorflow/python/ops/custom_gradient.py (hunk 1 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

File: tensorflow/python/ops/custom_gradient.py (hunk 2 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

File: tensorflow/python/ops/custom_gradient.py (hunk 3 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

File: tensorflow/python/ops/custom_gradient.py (hunk 4 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

File: tensorflow/python/ops/custom_gradient.py [hunk 8]
```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
...
+
+  ```python
+  >>> trainable = []
+  ... # List to store the trainable variables (weights, biases etc.)
+  ...
+  ... @tf.custom_gradient
+  ... def linear_poly(x):
+  ...   # Weights will be trainable parameter and will represent coefficients
+  ...   # of poly and initialized as ones.
+  ...   weights = tf.Variable(tf.ones([2]))
```
[the remark this answers is not in the document]
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

**Comment by ashutosh1919** [comment]
@allenlavoie , Thanks for feedback. I have modified the docstring as per your feedback. Please review.

File: tensorflow/python/ops/custom_gradient.py [hunk 9]
```diff
@@ -165,6 +165,73 @@ def second_order_and_transpose(ddy):
...
+    numpy=array([1., 1., 1.], dtype=float32)>
+  >>> tape.gradient(poly, weights)
+  <tf.Tensor: shape=(2,),
+    dtype=float32,
+    numpy=array([6., 3.], dtype=float32)>
+  ```
+
+  Above example illustrates two things:
+  - Usage of trainable variable `weights`.
+  - Computing its gradient.
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:226]: I would just take this summary out, I don't think it's adding much.

File: tensorflow/python/ops/custom_gradient.py [hunk 10]
```diff
@@ -165,6 +165,73 @@ def second_order_and_transpose(ddy):
...
+
+  In the example, the inner `grad_fn` accepts an extra `variables` input
+  parameter and also returns an extra `grad_vars` output. That extra argument
+  is passed if you have any trainable parameters in your model. You need to
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:230]: Missed this the first time, but I'd replace "if you have any trainable parameters in your model." with "if the forward function reads any variables."

File: tensorflow/python/ops/custom_gradient.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (allenlavoie) [reply tensorflow/python/ops/custom_gradient.py:186]: This still looks like it can be simplified: "assert variables is not None" and reduce the indent, no loop just "assert len(variables) == 1 and variables[0] is weights"

File: tensorflow/python/ops/custom_gradient.py [hunk 11]
```diff
@@ -165,6 +165,73 @@ def second_order_and_transpose(ddy):
...
+  ...     # dy/dx = weights[1] and we need to left multiply dpoly
+  ...     grad_xs = dpoly * weights[1]  # Scalar gradient
+  ...
+  ...     input_shape = len(x.shape)  # x contains how many dimensions
+  ...     grad_vars = []  # To store gradients of passed variables
+  ...     if variables is not None:
+  ...       for v in variables:
+  ...         if v is weights:
+  ...           # Manually computing dy/dweights
+  ...           dy_dw = tf.stack([x ** 1, x ** 0])
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:190]: Shouldn't this depend on dpoly?

File: tensorflow/python/ops/custom_gradient.py [hunk 12]
```diff
@@ -165,6 +165,73 @@ def second_order_and_transpose(ddy):
...
+  ...           if input_shape > 0:
+  ...             # Note here that reduce operation is performed
+  ...             # in axis so that it computes gradient of specific
+  ...             # example in batch and it doesn't reduce the entire batch
+  ...             # that is why we are only reducing on axis 1.
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:195]: Maybe to simplify the example just omit the batching bits and the `else:` and `tf.reduce_sum(tf.reshape(dy_dw, [2, -1]), axis=1)` instead

File: tensorflow/python/ops/custom_gradient.py (hunk 9 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py:226]: done

File: tensorflow/python/ops/custom_gradient.py (hunk 10 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py:230]: done

File: tensorflow/python/ops/custom_gradient.py (hunk 11 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

File: tensorflow/python/ops/custom_gradient.py (hunk 12 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

**Comment by ashutosh1919** [comment]
@allenlavoie , updated as per your feedback. Please review.

File: tensorflow/python/ops/custom_gradient.py [hunk 13]
```diff
@@ -165,6 +165,63 @@ def second_order_and_transpose(ddy):
...
+  ...     # dy/dx = weights[1] and we need to left multiply dpoly
+  ...     grad_xs = dpoly * weights[1]  # Scalar gradient
+  ...
+  ...     grad_vars = []  # To store gradients of passed variables
+  ...     assert variables is not None
+  ...     assert len(variables) == 1
+  ...     assert variables[0] is weights
+  ...     # Manually computing dy/dweights
+  ...     dy_dw = dpoly * tf.stack([x ** 1, x ** 0])
+  ...     # Note here that reduce operation is performed
```
Comment (allenlavoie) [inline tensorflow/python/ops/custom_gradient.py:190]: Maybe just remove this note? I'm not sure what it's saying. It's returning the correctly sized gradient for the example, so maybe no explanation is needed here.

**Review by allenlavoie — COMMENTED** [review]
One more nit. Looks good!

File: tensorflow/python/ops/custom_gradient.py (hunk 13 above)
Comment (ashutosh1919) [reply tensorflow/python/ops/custom_gradient.py]: done

**Comment by ashutosh1919** [comment]
@allenlavoie , i have updated commit. Can you please review again?

**Review by allenlavoie — APPROVED** [review]
Looks great!

**Comment by ashutosh1919** [comment]
@allenlavoie , Can you please approve again? It was failing doctest.

**Review by allenlavoie — APPROVED** [review]

**Comment by ashutosh1919** [comment]
@allenlavoie , extremely sorry. I don't know why i put ``` based docstring which was deprecated. I am not able to run doctests in my local that is why these problems. 
Can you please approve?

**Review by allenlavoie — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Comment by ashutosh1919** [comment]
@allenlavoie , @gbaned - This PR was approved 12 days ago but hasn't merged yet. Is there anything pending or failing for this?

**Comment by rthadur** [comment]
@ashutosh1919 there was some problems with our internal tool , we will process this soon. Thanks for your patience.

**Review by gbaned — APPROVED** [review]
